docs/contributing/999-rfcs/2024-06-02-ts-client.md

```diff
@@ -0,0 +1,354 @@
+The simulator has no API to add new permissions after the simulator has started and the scope of that feels too large for this RFC.
+
+As an MVP, the simulator can be configured to ignore permissions via `wing.toml`.
+Even though this is unsafe and cause cause local/cloud differences, wing is intentionally not claiming ownership of the client's host environment.
+
+### Type generation with `@lift`
+
+Wing's compiler has type information available for use, but only before preflight execution occurs.
+This makes an API like the following not feasible to implement using logic similar to the existing `extern` or `@inflight`:
+
+```wing
+client.lift("bucket", bucket, ["put"]);
+```
+
+While executing preflight, the knowledge about the type of the `bucket` instance is lost, so generating the actual client code will have no way to reference it.
+
+This RFC proposes changing the existing `lift` syntax as a way to capture the necessary type information during preflight.
+An important feature of this syntax is that it's entirely statically analyzable, so anything we need to do with it can be done before-preflight.
+The existing `lift` syntax is currently intended to be used only inside inflights to explicitly lift preflight data into the defined scope:
+
+```wing
+let b1 = new cloud.Bucket();
+inflight () => {
+  let var b = b1;
+  lift { b: [put] } {
+    b.put("key", "value");
+  }
+}
+```
+
+With some changes, with syntax could also be used as a way to create a preflight container for lifted data.
+
+```wing
+let b = new cloud.Bucket();
+let lifts = @lift({ b: [put] });
+//  ^^^^^ is a std.LiftQualificationsContainer
+```
+
+This is very similar to the `lift` mechanism in `@wingcloud/framework`
+
+```ts
+const b = new cloud.Bucket(app, "Bucket");
+const lifts = lift({ bucket }).grant({ bucket: ["put"] });
+```
+
+In the [RFC for explicit lift qualifications](https://github.com/winglang/wing/blob/main/docs/contributing/999-rfcs/2024-03-14-explicit-lift-qualification.md), there is a mention of a `std.LiftQualificationsContainer`.
+
+If this data structure is augmented to also include type information, it could be used to generate the client during preflight.
+The resulting data would look very similar to the existing _liftMap, with the addition of the link to the `.d.ts` generated by the compiler for the aggregate type information.
+
+### Client Contents
+
+The client package will contain the following:
+
+- package.json
+
+Minimal package.json with the name and version. For MVP, the package will not have any dependencies because the client code is self-contained.
+
+- index.js
+
+Exports the `Client` class. In this class constructor, the context is read from the environment variable `WING_CONTEXT` (or constructor arg) and is used to set the environment variables for the current process. Each lifted object will then be instantiated which utilizes the existing `_toInflight` logic and will consume all the necessary environment variables.
```

Comment:
> and is used to set the environment variables for the current process. 

This is an implementation detail. Worth mentioning that it might change in the future.

Reply:
This is indeed a section discussion implementation detail. I can add "for now"